rfcs/20191127-pip-structure.md

```diff
@@ -0,0 +1,212 @@
+
+Outline of the current structure:
+```
+tensorflow
+    __init__.py (contains "from tensorflow_core import *")
+
+tensorflow_core
+    python/...
+    lite/...
+    _api/v2
+        __init__.py
+        audio/__init__.py
+        autograph/__init__.py
+        ...
+```
+
+### Rationale behind current package structure
+#### Multiple version support
+To prepare for TensorFlow 2.0 launch, we added a way to build two versions: 1.x and 2.x. Each version has its own respective genrule that outputs file for 1.x or 2.x since API modules are different (for e.g. *tensorflow/manip/\_\_init\_\_.py* only exists in 1.x and not 2.x API). Now, bazel does not allow two genrules to output files to the same directory. Therefore, we have *_api/v1/* and *_api/v2/* subdirectories.
+
+Note that we could still place the API directly under *tensorflow/* in the pip package since a pip package contains a single version of TensorFlow. This option became out of reach when *tensorflow/contrib/lite/* was migrated to *tensorflow/lite/*. Now *tensorflow/lite/* API directory would conflict with *tensorflow/lite/* source directory if the API was under *tensorflow/* instead of *_api/vN/*.
+
+#### Circular dependencies
+Estimator depends on TensorFlow. At the same time, TensorFlow includes estimator as a part of its API. This creates a cycle.
+
+![alt_text](https://github.com/annarev/community/blob/pip_structure_rfc/rfcs/20191127-pip-structure/circular_dependency.png "Circular dependency
+between TensorFlow and Estimator.")
+
+#### Metapackage vs base package plans
+[Modular TensorFlow
+RFC](https://github.com/tensorflow/community/blob/master/rfcs/20190305-modular-tensorflow.md) proposes to keep two pip packages:
+tensorflow-base would only contain core TensorFlow (for e.g. no estimator).
+TensorFlow Metapackage would be a thin package defining composition of TensorFlow which includes base, estimator, keras and tensorboard.
+Note that this 2-package approach is not implemented yet. However, its proposal demonstrates how keeping a virtual pip package could be beneficial in the future.
+
+![alt_text](https://github.com/annarev/community/blob/pip_structure_rfc/rfcs/20191127-pip-structure/modular_structure.png "Proposed modular TensorFlow structure.")
+
+Current structure looks more like this (except *tensorflow/* and *tensorflow\_core/* are directories as opposed to separate pip packages) and meant to be the first step towards structure above:
+
+![alt_text](https://github.com/annarev/community/blob/pip_structure_rfc/rfcs/20191127-pip-structure/current_structure.png "Current TensorFlow structure.")
+
+### Current state of IDE code features
```

Comment:
For completeness, does it make sense to add a section can we add a section about CoLab and Jupyter? Many TF users use those.

Reply:
I don't think those are adversely affected by our package structure though

Reply:
Just checked `tf.keras.layers.*` tab completion in jupyter and you're correct that it's not affected.